crates/sui-types/src/messages.rs

```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
-                CallArg::Pure(_) => None,
-                CallArg::Object(ObjectArg::ImmOrOwnedObject(object_ref)) => {
-                    Some(vec![InputObjectKind::ImmOrOwnedMoveObject(*object_ref)])
-                }
-                CallArg::Object(ObjectArg::SharedObject {
-                    id,
-                    initial_shared_version,
-                    mutable,
-                }) => {
-                    let id = *id;
-                    let initial_shared_version = *initial_shared_version;
-                    let mutable = *mutable;
-                    Some(vec![InputObjectKind::SharedMoveObject {
-                        id,
-                        initial_shared_version,
-                        mutable,
-                    }])
-                }
-                CallArg::ObjVec(vec) => Some(
-                    vec.iter()
-                        .map(|obj_arg| match obj_arg {
-                            ObjectArg::ImmOrOwnedObject(object_ref) => {
-                                InputObjectKind::ImmOrOwnedMoveObject(*object_ref)
-                            }
-                            ObjectArg::SharedObject {
-                                id,
-                                initial_shared_version,
-                                mutable,
-                            } => {
-                                let id = *id;
-                                let initial_shared_version = *initial_shared_version;
-                                let mutable = *mutable;
-                                InputObjectKind::SharedMoveObject {
-                                    id,
-                                    initial_shared_version,
-                                    mutable,
-                                }
-                            }
-                        })
-                        .collect(),
-                ),
-            })
-            .flatten()
+            .flat_map(|arg| arg.input_objects())
             .chain([InputObjectKind::MovePackage(*package)])
             .collect()
     }
 }
 
+/// A series of commands where the results of one command can be used in future
+/// commands
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub struct ProgrammableTransaction {
+    /// Input objects or primitive values
+    pub inputs: Vec<CallArg>,
+    /// The commands to be executed sequentially. A failure in any command will
+    /// result in the failure of the entire transaction.
+    pub commands: Vec<Command>,
+}
+
+/// A single command in a programmable transaction.
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub enum Command {
+    /// A call to either an entry or a public Move function
+    MoveCall(Box<ProgrammableMoveCall>),
+    /// (Vec<forall T:key+store. T>, address)
+    /// It sends n-objects to the specified address. These objects must have store
+    /// (public transfer) and the previous owner must be an address.
```

Comment:
Re: "previous owner must be an address" is this true?  What about if the object was created fresh as a result of a previous move call?